pop warning message for cuda module when asan is built in

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35088 pop warning message for cuda module when asan is built in**

Pytorch cuda module is not compatible with asan built at this point. We will normally get an 'out of memory' error when doing cuda operation. Adding a warning message so people who is calling cuda operation knows why they get an unexpected 'out of memory' error.

Here is an example:

```
In [1]: import torch                                                                                                                                                                                                              

In [2]: a = torch.tensor([1,2], device='cuda')                                                                                                                                                                                    
THCudaCheck FAIL file=caffe2/aten/src/THC/THCGeneral.cpp line=47 error=2 : out of memory
Warning: torch.cuda: your pytorch binary has address sanitizer (asan) built in, asan is currently not compatible with torch.cuda module, you might get unexpected behavior (eg. out of memory, crash, etc.), please rebuild pytorch without asan if you need to use this module (THCPModule_initExtension at caffe2/torch/csrc/cuda/Module.cpp:431)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-153bb6e0562c> in <module>
----> 1 a = torch.tensor([1,2], device='cuda')

~/local/fbsource/fbcode/buck-out/dev/gen/pytorch/ifbpy#link-tree/torch/cuda/__init__.py in _lazy_init()
    151             raise AssertionError(
    152                 "libcudart functions unavailable. It looks like you have a broken build?")
--> 153         torch._C._cuda_init()
    154         # Some of the queued calls may reentrantly call _lazy_init();
    155         # we need to just return without initializing in that case.

RuntimeError: cuda runtime error (2) : out of memory at caffe2/aten/src/THC/THCGeneral.cpp:47
```

Differential Revision: [D20552708](https://our.internmc.facebook.com/intern/diff/D20552708)

Commit:
pop warning message for cuda module when asan is built in

[ghstack-poisoned]

Commit:
Update on "pop warning message for cuda module when asan is built in"


Differential Revision: [D20552708](https://our.internmc.facebook.com/intern/diff/D20552708)

[ghstack-poisoned]

gchanan commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -422,6 +422,15 @@ static void bindCudaDeviceProperties(PyObject* module) {
+#if __has_feature(address_sanitizer)
+  TORCH_WARN(
+    "torch.cuda: your pytorch binary has address sanitizer (asan) built in, "
+    "asan is currently not compatiable with torch.cuda module, "
```
nit: compatible

glaringlee replied:
updated

gchanan commented:
Echoing a chat I had with @colesbury:

Can you print the behavior of a program that fails in fbcode (and in OSS) where the warning comes up?  Might be worth considering making this an error if it always causes an error later anyway.

Commit:
Update on "pop warning message for cuda module when asan is built in"


Differential Revision: [D20552708](https://our.internmc.facebook.com/intern/diff/D20552708)

[ghstack-poisoned]

glaringlee commented:
Added an example in the description of this PR, showing when the warning message will be popped up.

gchanan reviewed (approved):
can you make sure it works in an OSS asan build before shipping it?

Commit:
Update on "pop warning message for cuda module when asan is built in"


Pytorch cuda module is not compatible with asan built at this point. We will normally get an 'out of memory' error when doing cuda operation. Adding a warning message so people who is calling cuda operation knows why they get an unexpected 'out of memory' error.

Here is an example:

```
In [1]: import torch                                                                                                                                                                                                              

In [2]: a = torch.tensor([1,2], device='cuda')                                                                                                                                                                                    
THCudaCheck FAIL file=caffe2/aten/src/THC/THCGeneral.cpp line=47 error=2 : out of memory
Warning: torch.cuda: your pytorch binary has address sanitizer (asan) built in, asan is currently not compatible with torch.cuda module, you might get unexpected behavior (eg. out of memory, crash, etc.), please rebuild pytorch without asan if you need to use this module (THCPModule_initExtension at caffe2/torch/csrc/cuda/Module.cpp:431)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-153bb6e0562c> in <module>
----> 1 a = torch.tensor([1,2], device='cuda')

~/local/fbsource/fbcode/buck-out/dev/gen/pytorch/ifbpy#link-tree/torch/cuda/__init__.py in _lazy_init()
    151             raise AssertionError(
    152                 "libcudart functions unavailable. It looks like you have a broken build?")
--> 153         torch._C._cuda_init()
    154         # Some of the queued calls may reentrantly call _lazy_init();
    155         # we need to just return without initializing in that case.

RuntimeError: cuda runtime error (2) : out of memory at caffe2/aten/src/THC/THCGeneral.cpp:47
```

Differential Revision: [D20552708](https://our.internmc.facebook.com/intern/diff/D20552708)

[ghstack-poisoned]

glaringlee commented:
Tested with CI pytorch_linux_xenial_py3_clang5_asan_build env as well, works fine.
